air/src/execution_step/execution_context/scalar_variables.rs

```diff
@@ -119,31 +164,34 @@ impl<'i> Scalars<'i> {
     }
 
     pub(crate) fn remove_iterable_value(&mut self, name: &str) {
-        self.iterable_values.remove(name);
+        self.iterable_variables.remove(name);
     }
 
     pub(crate) fn get_value(&'i self, name: &str) -> ExecutionResult<&'i ValueAggregate> {
-        self.values
+        self.non_iterable_variables
             .get(name)
-            .and_then(|scalars| {
-                scalars
-                    .iter()
-                    .take(self.fold_block_id + 1)
-                    .rev()
-                    .find_map(|scalar| scalar.as_ref())
+            .and_then(|values| {
+                let last_cell = values.last();
+                let value_not_invalidated = !self.invalidated_depths.contains(&last_cell.depth);
+
+                if is_global_value(last_cell.depth) || value_not_invalidated {
```

Comment:
I don't think that is_global_value is needed, because 0 will never be in invalidated_depths.